src/main/kotlin/com/ajax/cryptocurrency/service/CryptocurrencyService.kt

```diff
@@ -0,0 +1,65 @@
+package com.ajax.cryptocurrency.service
+
+import com.ajax.cryptocurrency.exception.CryptocurrencyPriceNotFoundException
+import com.ajax.cryptocurrency.exception.InvalidParamsForThePageException
+import com.ajax.cryptocurrency.model.Cryptocurrency
+import com.ajax.cryptocurrency.repository.CryptocurrencyRepository
+import org.springframework.data.domain.PageRequest
+import org.springframework.data.domain.Sort
+import org.springframework.stereotype.Service
+import java.io.File
+import java.io.FileWriter
+
+@Service
+class CryptocurrencyService(private val cryptocurrencyRepository: CryptocurrencyRepository) {
+    fun findMinMaxPriceByCryptocurrencyName(name: String, sortOrder: Int?): Cryptocurrency {
+        val unwrappedSortOrder = sortOrder ?: throw IllegalArgumentException("Sort order cannot be null")
+
+        val priceMinMax = cryptocurrencyRepository.findMinMaxByName(name, unwrappedSortOrder)
+            ?: throw CryptocurrencyPriceNotFoundException("Price of $name cryptocurrency not found")
+
+        return priceMinMax
+    }
+
+    fun findAll(): List<Cryptocurrency> {
+        return cryptocurrencyRepository.findAll()
+    }
+
+    fun getCryptocurrencyPages(name: String?, pageNo: Int, pageSize: Int): MutableList<Cryptocurrency?> {
+        when {
+            cryptocurrencyRepository.findAll()
+                .isEmpty() -> throw CryptocurrencyPriceNotFoundException("cryptocurrency prices not found")
+
+            pageNo < 0 || pageSize < 0 -> throw InvalidParamsForThePageException("invalid params for the page")
+            else -> {
+                val pageable = PageRequest.of(pageNo, pageSize, Sort.by("price"))
+                val pagedResult = name?.let {
+                    cryptocurrencyRepository.findCryptocurrencyPriceByCryptocurrencyName(name, pageable)
+                } ?: cryptocurrencyRepository.findAll(pageable)
+                return pagedResult.content
+            }
+        }
```

Comment:
It's bad idea to use `when` here. The code flow is confusing.

Suggestion to make the code more clear:
```kotlin
if (cryptocurrencyRepository.findAll().isEmpty()) { 
     // It's unclear for me why we should throw at all if there are not items in the DB
     // ¯\_(ツ)_/¯
     throw CryptocurrencyPriceNotFoundException("cryptocurrency prices not found")
}

val pageable = PageRequest.of(pageNo, pageSize, Sort.by("price"))
val pagedResult = if (name == null) {
    cryptocurrencyRepository.findAll(pageable)
} else {
    cryptocurrencyRepository.findCryptocurrencyPriceByCryptocurrencyName(name, pageable)
}
return pagedResult.content
```

Isn't it simpler to read?

Reply:
fixed